// @pixi/layout/dist/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const pixi_js = require("pixi.js");
const LayoutSystem = require("./core/LayoutSystem.cjs");
require("./core/mixins/ContainerMixin.cjs");
require("./core/mixins/TextMixin.cjs");
require("./core/mixins/ViewContainerMixin.cjs");
const DebugRenderer = require("./core/debug/DebugRenderer.cjs");
const Layout = require("./core/Layout.cjs");
const applyStyle = require("./core/style/applyStyle.cjs");
const formatStyles = require("./core/style/formatStyles.cjs");
const getNumberFromStyle = require("./core/utils/getNumberFromStyle.cjs");
const getPixiSize = require("./core/utils/getPixiSize.cjs");
const nearlyEqual = require("./core/utils/nearlyEqual.cjs");
const sortChildren = require("./core/utils/sort-children.cjs");
const yoga = require("./yoga.cjs");
pixi_js.extensions.add(LayoutSystem.LayoutSystem);
exports.LayoutSystem = LayoutSystem.LayoutSystem;
exports.DebugRenderer = DebugRenderer.DebugRenderer;
exports.Layout = Layout.Layout;
exports.applyStyle = applyStyle.applyStyle;
exports.formatStyles = formatStyles.formatStyles;
exports.getNumberFromStyle = getNumberFromStyle.getNumberFromStyle;
exports.getTypeFromStyle = getNumberFromStyle.getTypeFromStyle;
exports.getPixiSize = getPixiSize.getPixiSize;
exports.nearlyEqual = nearlyEqual.nearlyEqual;
exports.onChildAdded = sortChildren.onChildAdded;
exports.onChildRemoved = sortChildren.onChildRemoved;
exports.getYoga = yoga.getYoga;
exports.getYogaConfig = yoga.getYogaConfig;
exports.setYoga = yoga.setYoga;
exports.setYogaConfig = yoga.setYogaConfig;
