// @pixi/layout/dist/index.mjs
import { extensions } from "pixi.js";
import { LayoutSystem } from "./core/LayoutSystem.mjs";
import "./core/mixins/ContainerMixin.mjs";
import "./core/mixins/TextMixin.mjs";
import "./core/mixins/ViewContainerMixin.mjs";
import { DebugRenderer } from "./core/debug/DebugRenderer.mjs";
import { Layout } from "./core/Layout.mjs";
import { applyStyle } from "./core/style/applyStyle.mjs";
import { formatStyles } from "./core/style/formatStyles.mjs";
import { getNumberFromStyle, getTypeFromStyle } from "./core/utils/getNumberFromStyle.mjs";
import { getPixiSize } from "./core/utils/getPixiSize.mjs";
import { nearlyEqual } from "./core/utils/nearlyEqual.mjs";
import { onChildAdded, onChildRemoved } from "./core/utils/sort-children.mjs";
import { getYoga, getYogaConfig, setYoga, setYogaConfig } from "./yoga.mjs";
extensions.add(LayoutSystem);
export {
  DebugRenderer,
  Layout,
  LayoutSystem,
  applyStyle,
  formatStyles,
  getNumberFromStyle,
  getPixiSize,
  getTypeFromStyle,
  getYoga,
  getYogaConfig,
  nearlyEqual,
  onChildAdded,
  onChildRemoved,
  setYoga,
  setYogaConfig
};
