// @pixi/layout/dist/components/LayoutContainer.mjs
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
import { Container, Graphics as Graphics$1, Rectangle, Ticker } from "pixi.js";
import { Edge, BoxSizing } from "yoga-layout/load";
import { Trackpad } from "./trackpad/Trackpad.mjs";
function bindAndPreserve(base, source, methodNames) {
  const bound = {};
  const proto = Object.getPrototypeOf(base);
  for (const name of methodNames) {
    const method = proto[name];
    bound[name] = method.bind(base);
    base[name] = (...args) => source[name](...args);
  }
  return bound;
}
class LayoutContainer extends Container {
  constructor(params = {}) {
    const { layout, trackpad, background, children, ...options } = params;
    super(options);
    /** Graphics object used for rendering background and borders */
    __publicField(this, "background");
    __publicField(this, "stroke", new Graphics$1({ label: "stroke" }));
    /** The container that holds the overflow content */
    __publicField(this, "overflowContainer", new Container({
      label: "overflowContainer"
    }));
    /** Access to original Container methods */
    __publicField(this, "containerMethods");
    /** The trackpad for handling scrolling */
    __publicField(this, "_trackpad");
    /** Mask for overflow handling */
    __publicField(this, "_mask", new Graphics$1());
    /** Whether or not the background was created by the user */
    __publicField(this, "_isUserBackground", false);
    /** The maximum visual bounds of the layout */
    __publicField(this, "_visualBounds", new Rectangle());
    /** The scroll bounds for the container where scroll events can occur */
    __publicField(this, "_scrollBounds", new Rectangle());
    this.layout = layout ?? {};
    children == null ? void 0 : children.forEach((child) => this.addChild(child));
    this.overflowContainer.isOverflowContainer = true;
    this.background = background ?? new Graphics$1({ label: "background" });
    this._isUserBackground = !!background;
    this.addChild(this.background, this.overflowContainer, this._mask, this.stroke);
    this.containerMethods = bindAndPreserve(this, this.overflowContainer, [
      "addChild",
      "addChildAt",
      "removeChild",
      "removeChildAt",
      "getChildAt",
      "getChildIndex",
      "setChildIndex",
      "getChildByName",
      "removeChildren",
      "sortChildren",
      "swapChildren",
      "reparentChild",
      "reparentChildAt",
      "getChildByLabel",
      "getChildrenByLabel"
    ]);
    this._trackpad = new Trackpad({
      constrain: true,
      ...trackpad
    });
    this.eventMode = "static";
    this.on(
      "pointerdown",
      (e) => this.isPointWithinBounds(e.global.x, e.global.y) && this._trackpad.pointerDown(e.global)
    );
    this.on("pointerup", () => this._trackpad.pointerUp());
    this.on("pointerupoutside", () => this._trackpad.pointerUp());
    this.on(
      "pointermove",
      (e) => this.isPointWithinBounds(e.global.x, e.global.y) && this._trackpad.pointerMove(e.global)
    );
    this.on("pointercancel", () => this._trackpad.pointerUp());
    this.on("wheel", (e) => {
      var _a;
      const overflow = (_a = this.layout) == null ? void 0 : _a.style.overflow;
      if (overflow !== "scroll") {
        return;
      }
      if (!this.isPointWithinBounds(e.global.x, e.global.y)) {
        return;
      }
      const shift = e.shiftKey ? 1 : 0;
      const deltaX = e.deltaX * (shift ? 1 : -1);
      const deltaY = e.deltaY * (shift ? -1 : 1);
      const targetX = this._trackpad.xAxis.value - deltaX;
      const targetY = this._trackpad.yAxis.value - deltaY;
      if (this._trackpad.xAxis.constrainPercent >= 0) {
        this._trackpad.xAxis.value = Math.max(
          this._trackpad.xAxis.max,
          Math.min(this._trackpad.xAxis.min, targetX)
        );
      }
      if (this._trackpad.yAxis.constrainPercent >= 0) {
        this._trackpad.yAxis.value = Math.max(
          this._trackpad.yAxis.max,
          Math.min(this._trackpad.yAxis.min, targetY)
        );
      }
    });
    Ticker.shared.add(this.update, this);
  }
  /**
   * Computes the layout data for this container based on yoga calculations and draws the background.
   * @param computedLayout - The computed layout data from yoga
   * @returns Position and scale information for the container
   * @internal
   */
  computeLayoutData(computedLayout) {
    this._drawBackground(computedLayout);
    return {
      x: computedLayout.left,
      y: computedLayout.top,
      offsetX: 0,
      offsetY: 0,
      scaleX: 1,
      scaleY: 1
    };
  }
  /**
   * Updates the container mask based on overflow settings
   * @param width - Container width
   * @param height - Container height
   * @param radius - Border radius
   */
  _updateMask(width, height, radius = 0, borderWidth = 0, isContentBox = false) {
    const maskX = isContentBox ? -borderWidth : 0;
    const maskY = isContentBox ? -borderWidth : 0;
    const maskWidth = isContentBox ? width + borderWidth * 2 : width;
    const maskHeight = isContentBox ? height + borderWidth * 2 : height;
    this._mask.clear();
    this._mask.roundRect(0, 0, maskWidth, maskHeight, radius);
    this._mask.position.set(maskX, maskY);
    this._mask.fill(16777215);
    this._scrollBounds.set(
      maskX + borderWidth,
      maskY + borderWidth,
      maskWidth - borderWidth * 2,
      maskHeight - borderWidth * 2
    );
  }
  _updateBackground(computedLayout) {
    const layoutStyles = this.layout.style;
    const { backgroundColor, borderRadius } = layoutStyles;
    if (this._isUserBackground) {
      this.background.position.set(0, 0);
      this.background.setSize(computedLayout.width, computedLayout.height);
    } else {
      const background = this.background;
      background.clear();
      background.roundRect(0, 0, computedLayout.width, computedLayout.height, borderRadius ?? 0);
      if (backgroundColor != null) {
        background.fill({ color: backgroundColor });
      } else {
        background.fill({ color: 16777215, alpha: 0 });
      }
    }
  }
  /**
   * Draws the container's background including:
   * - Background color
   * - Border
   * - Border radius
   *
   * @param computedLayout - The computed layout data from yoga
   * @protected
   */
  _drawBackground(computedLayout) {
    var _a;
    let borderWidth = this.layout.yoga.getBorder(Edge.All);
    if (isNaN(borderWidth) || borderWidth < 0) {
      borderWidth = 0;
    }
    const boxSizing = this.layout.yoga.getBoxSizing();
    const alignment = boxSizing === BoxSizing.BorderBox ? 1 : 0;
    const layoutStyles = this.layout.style;
    const { borderColor, borderRadius } = layoutStyles;
    this._updateBackground(computedLayout);
    this.stroke.clear();
    if (borderWidth > 0 && borderColor != null) {
      this.stroke.roundRect(0, 0, computedLayout.width, computedLayout.height, borderRadius ?? 0);
      this.stroke.stroke({ color: borderColor, width: borderWidth, alignment });
    }
    const overflow = (_a = this.layout) == null ? void 0 : _a.style.overflow;
    if (overflow !== "visible") {
      this._updateMask(
        computedLayout.width,
        computedLayout.height,
        layoutStyles.borderRadius ?? 0,
        borderWidth,
        boxSizing === BoxSizing.ContentBox
      );
      this._updateScrollArea();
      this.setMask({ mask: this._mask });
      const maskWidth = computedLayout.width - this._visualBounds.width;
      const maskHeight = computedLayout.height - this._visualBounds.height;
      this._trackpad.xAxis.max = Math.min(0, maskWidth);
      this._trackpad.yAxis.max = Math.min(0, maskHeight);
    } else {
      this.mask = null;
      this._trackpad.xAxis.value = 0;
      this._trackpad.yAxis.value = 0;
      this.overflowContainer.position.set(0, 0);
    }
  }
  _updateScrollArea() {
    let maxBottom = 0;
    let maxRight = 0;
    for (let i = 0; i < this.layout.yoga.getChildCount(); i++) {
      const child = this.layout.yoga.getChild(i);
      const bottom = child.getComputedTop() + child.getComputedHeight();
      const right = child.getComputedLeft() + child.getComputedWidth();
      if (right > maxRight) maxRight = right;
      if (bottom > maxBottom) maxBottom = bottom;
    }
    maxRight += this.layout.yoga.getComputedPadding(Edge.Right);
    maxBottom += this.layout.yoga.getComputedPadding(Edge.Bottom);
    const borderWidth = this.layout.yoga.getBorder(Edge.All) || 0;
    maxRight += borderWidth;
    maxBottom += borderWidth;
    this._visualBounds.width = Math.max(0, maxRight);
    this._visualBounds.height = Math.max(0, maxBottom);
  }
  /**
   * Checks if a point is within the visual bounds of the container, accounting for world transformations.
   * @param x - The x-coordinate of the point to check
   * @param y - The y-coordinate of the point to check
   * @returns True if the point is within the visual bounds, false otherwise
   */
  isPointWithinBounds(x, y) {
    return this._scrollBounds.contains(x - this.worldTransform.tx, y - this.worldTransform.ty);
  }
  update() {
    var _a;
    const overflow = (_a = this.layout) == null ? void 0 : _a.style.overflow;
    if (overflow !== "scroll") {
      return;
    }
    this._trackpad.update();
    this.overflowContainer.x = this._trackpad.x;
    this.overflowContainer.y = this._trackpad.y;
  }
  destroy(options) {
    super.destroy(options);
    Ticker.shared.remove(this.update, this);
  }
}
export {
  LayoutContainer
};
