// lzma1/lib/range-encoder.js
export class RangeEncoder {
    stream = null;
    low = 0n;
    rrange = 0;
    cache = 0;
    cacheSize = 0;
    position = 0n;
    constructor() {
        // Initialize with default values
    }
    /**
     * Set output stream for encoding
     */
    setStream(stream) {
        this.stream = stream;
    }
    /**
     * Initialize range encoder
     */
    init() {
        this.position = 0n;
        this.low = 0n;
        this.rrange = -1;
        this.cacheSize = 1;
        this.cache = 0;
    }
    /**
     * Encode a single bit using probability model
     */
    encodeBit(probs, index, bit) {
        const prob = probs[index];
        const newBound = (this.rrange >>> 11) * prob;
        if (bit === 0) {
            this.rrange = newBound;
            probs[index] = prob + ((2048 - prob) >>> 5);
        }
        else {
            this.low += BigInt(newBound);
            this.rrange -= newBound;
            probs[index] = prob - (prob >>> 5);
        }
        if (!(this.rrange & -0x1000000)) {
            this.rrange <<= 8;
            this.shiftLow();
        }
    }
    /**
     * Encode direct bits (without probability model)
     */
    encodeDirectBits(value, numTotalBits) {
        for (let i = numTotalBits - 1; i >= 0; i--) {
            this.rrange >>>= 1;
            if (((value >>> i) & 1) === 1) {
                this.low += BigInt(this.rrange);
            }
            if (!(this.rrange & -0x1000000)) {
                this.rrange <<= 8;
                this.shiftLow();
            }
        }
    }
    /**
     * Encode bit tree symbol
     */
    encodeBitTree(numBitLevels, models, symbol, startIndex = 1) {
        let m = 1;
        for (let bitIndex = numBitLevels; bitIndex !== 0; bitIndex--) {
            const bit = (symbol >>> (bitIndex - 1)) & 1;
            this.encodeBit(models, startIndex + m - 1, bit);
            m = (m << 1) | bit;
        }
    }
    /**
     * Reverse encode bits
     */
    reverseEncodeBits(numBitLevels, models, symbol, startIndex) {
        let m = 1;
        for (let i = 0; i < numBitLevels; i++) {
            const bit = symbol & 1;
            this.encodeBit(models, startIndex + m, bit);
            m = (m << 1) | bit;
            symbol >>>= 1;
        }
    }
    /**
     * Finish encoding and flush remaining data
     */
    finish() {
        for (let i = 0; i < 5; i++) {
            this.shiftLow();
        }
    }
    /**
     * Shift low value and write to stream
     */
    shiftLow() {
        const lowHi = Number((this.low >> 32n) & 0xffffffffn);
        if (lowHi !== 0 || Number(this.low & 0xffffffffn) < 0xFF000000) {
            this.position += BigInt(this.cacheSize);
            let temp = this.cache;
            do {
                this.writeToStream(temp + lowHi);
                temp = 255;
            } while (--this.cacheSize !== 0);
            this.cache = (Number(this.low & 0xffffffffn) >>> 24) & 0xFF;
        }
        this.cacheSize++;
        this.low = BigInt(Number(this.low & 0xffffffffn) & 0xFFFFFF) << 8n;
    }
    /**
     * Write byte to output stream
     */
    writeToStream(value) {
        if (!this.stream) {
            return;
        }
        this.stream.writeByte(value & 0xFF);
    }
}
