// lzma1/lib/streams.js
/**
 * Read-only input buffer wrapping a Uint8Array.
 */
export class InputBuffer {
    buf;
    pos = 0;
    count;
    constructor(data) {
        this.buf = data;
        this.count = data.length;
    }
    readByte() {
        if (this.pos >= this.count)
            return -1;
        return this.buf[this.pos++] & 0xFF;
    }
    readBytes(dest, off, len) {
        if (this.pos >= this.count)
            return -1;
        len = Math.min(len, this.count - this.pos);
        dest.set(this.buf.subarray(this.pos, this.pos + len), off);
        this.pos += len;
        return len;
    }
    get remaining() {
        return this.count - this.pos;
    }
}
/**
 * Growable output buffer backed by a Uint8Array.
 */
export class OutputBuffer {
    buf;
    count = 0;
    constructor(capacity = 32) {
        this.buf = new Uint8Array(capacity);
    }
    grow(requiredSize) {
        const newSize = Math.max(this.buf.length * 2, requiredSize);
        const newBuf = new Uint8Array(newSize);
        newBuf.set(this.buf.subarray(0, this.count));
        this.buf = newBuf;
    }
    writeByte(b) {
        if (this.count >= this.buf.length) {
            this.grow(this.count + 1);
        }
        this.buf[this.count++] = b;
    }
    writeBytes(src, off, len) {
        const requiredSize = this.count + len;
        if (requiredSize > this.buf.length) {
            this.grow(requiredSize);
        }
        this.buf.set(src.subarray(off, off + len), this.count);
        this.count += len;
    }
    write(buf) {
        this.writeBytes(buf, 0, buf.length);
    }
    toArray() {
        return this.buf.slice(0, this.count);
    }
}
